ext/MitsubaLoader/CMitsubaLoader.h

```diff
@@ -49,269 +49,165 @@ namespace bsdf
 	_IRR_STATIC_INLINE_CONSTEXPR uint32_t BITFIELDS_SHIFT_OPACITY_TEX = 10u;
 	_IRR_STATIC_INLINE_CONSTEXPR uint32_t BITFIELDS_MASK_PHONG_EXP_TEX = 0x1u;
 	_IRR_STATIC_INLINE_CONSTEXPR uint32_t BITFIELDS_SHIFT_PHONG_EXP_TEX = 16u;
-	_IRR_STATIC_INLINE_CONSTEXPR uint32_t BITFIELDS_MASK_MIXTURE_CHILD_COUNT = 0x1fu;
-	_IRR_STATIC_INLINE_CONSTEXPR uint32_t BITFIELDS_SHIFT_MIXTURE_CHILD_COUNT = 10u;
+
+	//padding which make sure struct has sizeof at least min_desired_size but aligned to 4*sizeof(uint32_t)=16
+	//both size and desired_size are meant to be expressed in 4 byte units (i.e. to express size of 8, `size` should be 2)
+	template<size_t size, size_t min_desired_size, size_t diff = (min_desired_size-size)>
+	using padding_t = uint32_t[ ((min_desired_size + 3ull) & ~(3ull)) - size ];
+	//in 4 byte units
+	constexpr size_t max_bsdf_struct_size = 10ull;
+
 #include "irr/irrpack.h"
-	/**
-	GLSL:
-	struct {
-		uint bitfields;
-		vec2 alpha;
-		vec3 reflectance;
-	};
-	*/
 	struct alignas(16) SAllDiffuse
 	{
 		uint32_t bitfields;
-		float dummy;
 		//if flag decides to use alpha texture, {alpha[0..1]} is bindless texture ID (bit-cast to uvec2)
 		//otherwise alpha[0] is single-float alpha
-		float alpha[2];
-		//if flag decides to use reflectance texture, {reflectance[0..1]} is bindless texture ID (bit-cast to uvec2)
-		//otherwise reflectance[0..2] is constant reflectance
-		float reflectance[3];
+		uint32_t alpha[2];
+		//if flag decides to use reflectance texture, `reflectance` is bindless texture ID (bit-cast to uvec2)
+		//otherwise `reflectance` is constant reflectance in RGB19E7 format
+		uint64_t reflectance;
+		padding_t<5, max_bsdf_struct_size> padding;
```

Comment:
that will backfire since vec2 (`reflectance`) cannot straddle a `vec4` boundary

Reply:
I removed all the paddings because GLSL representation will be a struct of a few uvec4s as you suggested. So that everything is packed as possible. I just thought wouldn't it be better if one member didn't lie in just one uvec4 (not in between since I need >1 fetches then).

>I thought roughness could be a texture as well as a float?

yes, single-floats are uint32[2] for my convenience to easier access lower 32 bits (where the float is stored)

Reply:
> yes, single-floats are uint32[2] for my convenience to easier access lower 32 bits (where the float is stored)

Good idea.